src/common/options.cc

```diff
@@ -8589,8 +8589,15 @@ std::vector<Option> get_mds_client_options() {
     .set_default(2)
     .set_min(1)
     .set_description("Size of thread pool for ASIO completions")
+    .add_tag("client"),
+
+    Option("client_shutdown_timeout", Option::TYPE_SECS, Option::LEVEL_ADVANCED)
+    .set_default(30)
```

Comment:
Is 30s enough? Suppose I have a network hiccup and a bunch of outstanding data  to write back. Might this lead to data corruption when writeback occurs? Note that the client may not end up blacklisted in this case, as 30s is not long enough for that to occur.

I think the default probably ought to be on the order of the auto-blacklisting timeout in the MDS. IOW, If you're going to end up blacklisted anyway then you might as well just give up. Otherwise, you probably want to wait for the MDS to come back.

Reply:
Or just default to how it is now which is indefinite. Is there a similar config in kclient?

Before closing sessions, clients should have flushed buffered data, caps and synchronously wait for those flushes to get acked (on unmount). That should be enough to not run into any kind of data corruption.

Reply:
I'm not aware of a similar setting in the kclient. If the data got written back, then that's probably ok for regular files. Could we end up waiting here while there are still waiting on async dirops to complete?

Also, how often do we end up in a situation where the only thing we're hung on is the session close request? I do wonder if we might plumb this in only to find that we end up similarly hung trying to send a MDS request prior to this point when umounting.

Reply:
> I'm not aware of a similar setting in the kclient. If the data got written back, then that's probably ok for regular files. Could we end up waiting here while there are still waiting on async dirops to complete?

This is called via unmount or client calling shutdown directly. Calling shutdown directly without flushing buffered data is surely problematic and there are no checks in shutdown to catch this. For the normal case of this getting invoke via unmount -- that should have taken care of any async ops I guess...

> 
> Also, how often do we end up in a situation where the only thing we're hung on is the session close request? I do wonder if we might plumb this in only to find that we end up similarly hung trying to send a MDS request prior to this point when umounting.

Not often. The sequence that was seen in teuthology tests:
 
- Client does a mount -- libcephfs picks up rank 0 from mdsmap and initiates session open
- mds rank 0 goes offline
- mount timeout kicks in
- shutdown is initiated -- normally there is no need of shutdown when mount did not go through, however, shutdown initiates session close even if the filesystem is not mounted since sessions might exist to send MDS commands: https://github.com/ceph/ceph/blob/master/src/client/Client.cc#L565

(Note that I haven't handled the case when MDS is back up online before timeout)

Reply:
It sounds reasonable to me. In the future, we can use this same config to broaden the scope so that we catch MDS/OSDs becoming unavailable during a `umount`.

Reply:
> Hmm ok. This sounds reasonably harmless then. If this is unlikely to have any effect on normal operation, maybe we should just dispense with the tunable and hardcode a timeout?

Timeout should be configurable IMO.

Reply:
> It sounds reasonable to me. In the future, we can use this same config to broaden the scope so that we catch MDS/OSDs becoming unavailable during a `umount`.

@batrick I think we should keep the default as 0 (no timeout, current behaviour).